File: Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int m = matrix.size();
        if (m == 0) return false;
        int n = matrix[0].size();
        int low = 0, high = m*n - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            int cur = matrix[mid/n][mid%n];
            if (cur == target) {
                return true;
            } else if (cur > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        return false;
    }
};

File: Search a 2D Matrix.cs
public class Solution {
    public bool SearchMatrix(int[,] matrix, int target) {
        if (matrix == null || matrix.Length == 0) {
            return false;
        }
        
        var low = 0;
        var high = matrix.Length - 1;
        while (low <= high) {
            var mid = low + (high - low) / 2;
            var i = mid / matrix.GetLength(1);
            var j = mid % matrix.GetLength(1);
            if (matrix[i, j] == target) {
                return true;
            } else if (matrix[i, j] > target) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        return false;
    }
}
